src/astral/data-utils.js

```diff
@@ -95,6 +95,51 @@ const getCharacter = async (name) => {
     return char ? char.id : undefined;
 }
 
+const getCharacterData = async (id) => await (await fetch(
+    `${location.origin}/api/game/${getRoom()}/character/${id}`, 
+    {
+        method: "GET", 
+        headers: await getHeaders()
+    }
+)).json();
+
+const updateCustomAttribute = async (characterData, attrbiuteName, value) => {
+    let attr = characterData.customAttributes.find(attr => attr.name == attrbiuteName);
+    if (!attr) {
+        attr = {
+            name: attrbiuteName,
+            value: value.toString()
+        }
+        characterData.customAttributes.push(attr);
+        return;
+    }
+    attr.value = value.toString();
+}
+
+const updateResourceBar = async (characterData, barName, value, maxValue, color = null, actualValue = null, actualMaxValue = null, display = true) => {
+    let bar = characterData.resourceBars.find(attr => attr.label == barName);
```

Comment:
any possibility the character data doesn't have a `resourceBars` field? same for `customAttributes` above? Try a blank new character and make sure those two fields are defined and of type array.

Reply:
From what I know they should all have those attributes since even new tokens have hp and a resourceBar, but will add some sanity checks.

Reply:
On second thought if they were undefined it would be better to simply crash and let the catch below throw the log. You can't do partial updates to `resourceBars` and `customAttributes` at least from what I have tried without breaking the actor/character. You need to send all existing bars and attributes. So if we were to populate them if undefined for whatever reason, it might cause problems down the line by maybe removing attributes created in the stat sheet by the user.